packages/loopring_v3/contracts/amm/libamm/AmmExitProcess.sol

```diff
@@ -79,93 +38,152 @@ library AmmExitProcess
         )
         internal
     {
-        S.validatePoolTransaction(
+        S.checkPoolTxApproval(
             exit.owner,
-            AmmExitRequest.hashPoolExit(ctx.domainSeparator, exit),
+            AmmExitRequest.hash(ctx.domainSeparator, exit),
             signature
         );
 
-        if (signature.length == 0) {
-            // This is an onchain exit, we're processing it now so stop tracking it.
-            S.isExiting[exit.owner] = false;
+        bool burnFromLayer1 = (
+            exit.direction == AmmData.Direction.L1_TO_L1 ||
+            exit.direction == AmmData.Direction.L1_TO_L2
+        );
+
+        if (burnFromLayer1) {
+            require(signature.length == 0, "LAYER1_EXIT_WITH_OFFCHAIN_APPROVAL_DISALLOWED");
+
+            delete S.exitLockIdx[msg.sender];
+            delete S.exitLocks[S.exitLocksIndex];
+            S.exitLocksIndex++;
```

Comment:
`msg.sender` -> `exit.owner`?

This data is deleted but there doesn't seem to be any link with the actual exit being processed? Also don't understand what the purpose it of this data. Maybe for manual withdrawals when the pool is shutdown is possible just be posting the request data onchain again. Still not as user friendly as just a single balance like it was where people could in theory do it themselves by interacting with the contract with a simple `withdraw`, but still very doable because it's just reposting data from events (seems easier than extra onchain data).

Reply:
The `delete S.exitLocks[S.exitLocksIndex];` doesn't seem to be directly related to the exit being processed. I think for the current system `S.exitLockIdx[exit.owner] == S.exitLocks[S.exitLocksIndex]` needs to be true, otherwise it's possible an unrelated `exitLocks` is deleted when the exits are done out-of-order.